pkg/shardddl/optimism/keeper.go

```diff
@@ -51,15 +51,23 @@ func (lk *LockKeeper) RebuildLocksAndTables(
 		lock *Lock
 		ok   bool
 	)
-	for _, taskInfos := range ifm {
-		for _, sourceInfos := range taskInfos {
-			for _, schemaInfos := range sourceInfos {
-				for _, info := range schemaInfos {
+	for task, taskInfos := range ifm {
+		for source, sourceInfos := range taskInfos {
+			for schema, schemaInfos := range sourceInfos {
+				for table, info := range schemaInfos {
 					lockID := utils.GenDDLLockID(info.Task, info.DownSchema, info.DownTable)
 					if lock, ok = lk.locks[lockID]; !ok {
-						lk.locks[lockID] = NewLock(cli, lockID, info.Task, info.DownSchema, info.DownTable, lockJoined[lockID], lockTTS[lockID])
-						lock = lk.locks[lockID]
+						lock = NewLock(cli, lockID, info.Task, info.DownSchema, info.DownTable, lockJoined[lockID], lockTTS[lockID])
 					}
+					// filter info which doesn't have SourceTable
+					// SourceTable will be changed after user update block-allow-list
```

Comment:
as opposed to cleaning up error info here
maybe we can celar outated info in etcd after user update `block-allow-list` is a better idea?

Reply:
Not reliable in current design.
```
user do a ddl, 
stop-task
drop table
redo task(start-task remove meta)
```
the ddl info still in etcd, but sourceTable was deleted.

Reply:
can we clear data when `remove-meta`?

Reply:
It's a idea